app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,365 @@
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          toggleHotspot();
+        } else {
+          //TO DO: show Dialog() + within that add check mobile Data check later.
+          //if (isMobileDataEnabled(context)) {
+          //  mobileDataDialog();
+          //} else {
+          if (flag) {
+            serverStopped();
+            isStarted = false;
+          } else {
+            startHotspotDialog();
+          }
+          //}
+        }
+      }
+    });
+  }
+
+  @RequiresApi(api = Build.VERSION_CODES.O)
+  void toggleHotspot() {
+    //Check if location permissions are granted
+    if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
+        == PackageManager.PERMISSION_GRANTED) {
+      if (checkHotspotState(this)) //If hotspot is already enabled, turn it off
+      {
+        startService(ACTION_TURN_OFF_AFTER_O);
+      } else //If hotspot is not already enabled, then turn it on.
+      {
+        setupLocationServices();
+      }
+    } else {
+      //Ask location permission if not granted
+      ActivityCompat.requestPermissions(this,
+          new String[] { Manifest.permission.ACCESS_FINE_LOCATION },
+          MY_PERMISSIONS_ACCESS_FINE_LOCATION);
+    }
+  }
+
+  // This method checks if mobile data is enabled in user's device.
+  static boolean isMobileDataEnabled(Context context) {
+    boolean enabled = false;
+    ConnectivityManager cm =
+        (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+    try {
+      Class cmClass = Class.forName(cm.getClass().getName());
+      Method method = cmClass.getDeclaredMethod("getMobileDataEnabled");
+      method.setAccessible(true);
+      enabled = (Boolean) method.invoke(cm);
+    } catch (Exception e) {
+      Log.e("TAG", e.toString());
+    }
+    return enabled;
+  }
+
+  //This method sends the user to data usage summary settings activity
+  private void disableMobileData() {
+    Intent intent = new Intent();
+    intent.setComponent(new ComponentName("com.android.settings",
+        "com.android.settings.Settings$DataUsageSummaryActivity"));
+    startActivity(intent);
+  }
+
+  @Override public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+      @NonNull int[] grantResults) {
+    switch (requestCode) {
+      case MY_PERMISSIONS_ACCESS_FINE_LOCATION: {
+        if (grantResults.length > 0
+            && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            toggleHotspot();
+          }
+        }
+        break;
+      }
+      default:
+        break;
+    }
+  }
+
+  @Override public boolean onKeyDown(int keyCode, KeyEvent event) {
+    if (event.getAction() == KeyEvent.ACTION_DOWN) {
+      switch (keyCode) {
+        case KeyEvent.KEYCODE_BACK:
+          //Code for webserver
+          if (isStarted) {
+            new android.app.AlertDialog.Builder(this)
```

Comment:
Dialogs should be shown by AlertDialogShower and these strings should be in strings.xml